kotlinx-coroutines-core/native/src/MultithreadedDispatchers.kt

```diff
@@ -79,14 +79,23 @@ private class MultiWorkerDispatcher(
         }
     }
 
-    private fun workerRunLoop() = runBlocking {
-        // NB: we leverage tail-call optimization in this loop, do not replace it with
-        // .receive() without proper evaluation
-        for (task in tasksQueue) {
-            /**
-             * Any unhandled exception here will pass through worker's boundary and will be properly reported.
-             */
-            task.run()
+    private fun workerRunLoop() {
+        try {
+            runBlocking {
+                // NB: we leverage tail-call optimization in this loop, do not replace it with
+                // .receive() without proper evaluation
+                for (task in tasksQueue) {
+                    /**
+                     * Any unhandled exception here will pass through worker's boundary and will be properly reported.
+                     */
+                    task.run()
+                }
+            }
+        } catch (e: IllegalStateException) {
+            // We deliberately ignore ISE here as they are expected
+            // due to peculiarities of Workers API.
+            // Unfortunately, race-free termination is unachievable by its current state,
+            // see https://github.com/Kotlin/kotlinx.coroutines/issues/3578
```

Comment:
Even if we decide to use https://github.com/Kotlin/kotlinx.coroutines/pull/3585, let's still use this change. It prevents `IllegalStateException` in the case when the worker is terminated before ever entering `runBlocking`.

Reply:
Now that I think about it, it doesn't look like `IllegalStateException` can really happen anymore.
```kotlin
    @Test
    fun testRunBlockingInAlreadyTerminatedWorker() {
        val finished = Channel<Unit>()
        Worker.start().apply {
            executeAfter(1_000_000) {
                runBlocking {
                    finished.send(Unit)
                }
            }
        }.requestTermination()
        runBlocking {
            finished.receive()
        }
    }
```
It passes with no problems. When the worker starts `runBlocking`, it's obviously doing some work, and that's enough to prevent it from terminating. So, it looks like the issue has been mitigated already.